The NHS dm+d has a flag to indicate as to whether this item relates to a pack that is only to be made available through
hospital prescribing [(see tech spec of dm+d)](https://www.nhsbsa.nhs.uk/sites/default/files/2017-02/Data_Model_R2_v3.1_May_2015.pdf). 

Unusually the information appears to be held at AMP and  AMPP level. It is held in "Restrictions on availability" at AMP level and in the "product prescribing information" at AMPP level. This notebook explores what prescribing is going on in primary care for the possibility of a new measure.

In [1]:
##importing libraries
import pandas as pd
import numpy as np
from ebmdatalab import bq, maps, charts
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec

In [2]:
##ensure £ and pence are set right
pd.set_option('display.float_format', lambda x: '%.2f' % x)

### AMP LEVEL

In [3]:
sql = '''
SELECT
  PCO_Name,
  PCO_Code,
  Practice_Name, 
  Practice_Code,
  presc.bnf_code,
  dmd.id,
  avail_restrict,
  bnf_description AS bnf_name,
  Items,
  Quantity,
  Actual_Cost
FROM
  ebmdatalab.tmp_eu.raw_prescribing_data_2019_04 AS presc ##this is raw quantity file. using this to see if there are patterns of small quantities
Left JOIN  
  ebmdatalab.dmd.amp AS dmd
ON
presc.bnf_code = dmd.bnf_code
WHERE
avail_restrict = 8 ## this is code for hospital Rx
'''

df_amp = bq.cached_read(sql, csv_path='hospamp.csv')
df_amp.rename(columns={'PCO_Code':'pct'}, inplace=True)
df_amp.head(5)

C:\Users\bmackenna\Anaconda3\lib\site-packages\pandas\io\gbq.py:149: FutureWarning: verbose is deprecated and will be removed in a future version. Set logging level in order to vary verbosity
  credentials=credentials, verbose=verbose, private_key=private_key)


PCO_Name    pct           Practice_Name Practice_Code  \
0  EASTERN CHESHIRE CCG  01C00  MIDDLEWOOD PARTNERSHIP        N81022   
1  EASTERN CHESHIRE CCG  01C00  MIDDLEWOOD PARTNERSHIP        N81022   
2  EASTERN CHESHIRE CCG  01C00  MIDDLEWOOD PARTNERSHIP        N81022   
3  EASTERN CHESHIRE CCG  01C00  MIDDLEWOOD PARTNERSHIP        N81022   
4  EASTERN CHESHIRE CCG  01C00  MIDDLEWOOD PARTNERSHIP        N81022   

          bnf_code                 id  avail_restrict  \
0  0211000P0AAAAAA  22451411000001103               8   
1  0211000P0AAAAAA  22786211000001105               8   
2  0211000P0AAAAAA  23268711000001101               8   
3  0211000P0AAAAAA  34756011000001103               8   
4  0211000P0AAAAAA  35750911000001105               8   

                               bnf_name  Items  Quantity  Actual_Cost  
0  Tranexamic Acid_Inj 100mg/ml 5ml Amp      2         5        13.96  
1  Tranexamic Acid_Inj 100mg/ml 5ml Amp      2         5        13.96  
2  Tranexamic Acid_Inj 100mg/ml 5ml Amp      2         5        13.96  
3  Tranexamic Acid_Inj 100mg/ml 5ml Amp      2         5        13.96  
4  Tranexamic Acid_Inj 100mg/ml 5ml Amp      2         5        13.96

In [4]:
#monthly total
df_amp.sum()[['Items','Actual_Cost']]

Items              1568
Actual_Cost   115078.41
dtype: object

In [5]:
##groupby bnf name  to see "biggest offenders" in terms of items
df_amp_product = df_amp.groupby(['id', 'bnf_name', 'avail_restrict']).sum().sort_values(by = 'Items', ascending = False)
df_amp_product

Items  \
id                bnf_name                                 avail_restrict          
17652211000001105 Colistimeth Sod_Inj 1,000,000u Vl        8                 385   
27953111000001108 Midazolam HCl_Inj 2mg/ml 5ml Amp         8                 178   
22786211000001105 Tranexamic Acid_Inj 100mg/ml 5ml Amp     8                  94   
22451411000001103 Tranexamic Acid_Inj 100mg/ml 5ml Amp     8                  94   
23268711000001101 Tranexamic Acid_Inj 100mg/ml 5ml Amp     8                  94   
35750911000001105 Tranexamic Acid_Inj 100mg/ml 5ml Amp     8                  94   
34756011000001103 Tranexamic Acid_Inj 100mg/ml 5ml Amp     8                  94   
11260411000001104 Viridal Duo_S/Pack Inj 10mcg Cart        8                  62   
3998211000001106  Gentamicin Sulf_Inj 40mg/ml 2ml Vl       8                  44   
11260711000001105 Viridal Duo_S/Pack Inj 20mcg Cart        8                  33   
32711911000001108 Levetiracetam_I/V Inf 100mg/ml 5ml Vl    8                  33   
34754411000001104 Levetiracetam_I/V Inf 100mg/ml 5ml Vl    8                  33   
34751011000001105 Ceftriaxone Sod_Inj 1g Vl                8                  32   
11261011000001104 Viridal Duo_S/Pack Inj 40mcg Cart        8                  31   
34751211000001100 Ceftriaxone Sod_Inj 2g Vl                8                  30   
34551211000001101 Ketamine_Inj 50mg/ml 10ml Vl             8                  24   
19572711000001100 Ketorolac Trometamol_Inj 30mg/ml 1ml Amp 8                  16   
33899511000001107 Ketorolac Trometamol_Inj 30mg/ml 1ml Amp 8                  16   
31685111000001103 Omeprazole_I/V Inf 40mg Vl               8                  13   
34754811000001102 Omeprazole_I/V Inf 40mg Vl               8                  13   
35775411000001106 Omeprazole_I/V Inf 40mg Vl               8                  13   
10423311000001100 Phenylbutazone_Tab 100mg                 8                  12   
34748211000001106 Amoxicillin Sod_Inj 500mg Vl             8                  11   
34755011000001107 Pantoprazole_Inj 40mg Vl                 8                   9   
35306811000001105 Meropenem_Inj 500mg Vl (Dry)             8                   8   
4681611000001103  Chirocaine_Inj 2.5mg/ml 10ml Amp         8                   8   
34753011000001106 Fluclox Sod_Inj 250mg Vl                 8                   6   
34752411000001106 Co-Amoxiclav_Inj 1g/200mg Vl             8                   5   
34750611000001108 Ceftazidime Pentahyd_Inj 2g Vl           8                   5   
32391011000001106 Meropenem_Inj 1g Vl (Dry)                8                   5   
...                                                                          ...   
28962211000001109 Zoledronic Acid_I/V Inf 800mcg/ml 5ml Vl 8                   4   
4213611000001104  Vancomycin HCl_I/V Inf 500mg Vl          8                   4   
34752811000001108 Fluclox Sod_Inj 1g Vl                    8                   4   
34758311000001101 Vancomycin HCl_I/V Inf 500mg Vl          8                   4   
34930711000001106 Emtricitabine/Tenofovir_Tab 200mg/245mg  8                   3   
34755411000001103 Piperacillin/Tazobactam_I/VInf 4g/0.5gVl 8                   3   
4658711000001106  Promixin_Inj 1,000,000u Vl (Dry)         8                   3   
34753211000001101 Fluclox Sod_Inj 500mg Vl                 8                   3   
3614611000001101  Methotrexate_Inj 2.5mg/ml 2ml Vl         8                   3   
34750811000001107 Ceftriaxone Sod_Inj 250mg Vl             8                   3   
15980311000001108 Enbrel_Inj 50mg/ml 1ml Pf MyClic Pen     8                   3   
15442811000001106 Piperacillin/Tazobactam_I/VInf 4g/0.5gVl 8                   3   
21291311000001103 Gelaspan_Ecobag I/V Inf 4% 500ml         8                   2   
34750411000001105 Ceftazidime Pentahyd_Inj 1g Vl           8                   2   
34750011000001101 Cefotaxime Sod_Inj 500mg Vl              8                   2   
34749811000001100 Cefotaxime Sod_Inj 1g Vl 

In [6]:
sql = '''
SELECT
  PCO_Name,
  PCO_Code,
  Practice_Name, 
  Practice_Code,
  presc.bnf_code,
  dmdone.id,
  dmdtwo.hosp,
  bnf_description AS bnf_name,
  Items,
  Quantity,
  Actual_Cost
FROM
  ebmdatalab.tmp_eu.raw_prescribing_data_2019_04 AS presc ##this is raw quantity file. using this to see if there are patterns of small quantities
Left JOIN  
  ebmdatalab.dmd.ampp AS dmdone ##needed to link ampp code to bnf code
ON
presc.bnf_code = dmdone.bnf_code
LEFT JOIN
ebmdatalab.dmd.prescribinfo AS dmdtwo ##this is where product prescribing info is held
ON
dmdone.id = dmdtwo.ampp 
WHERE
dmdtwo.hosp = true
'''

df_ampp = bq.cached_read(sql, csv_path='hospampp.csv')
df_ampp.rename(columns={'PCO_Code':'pct'}, inplace=True)
df_ampp.head(5)

PCO_Name    pct               Practice_Name Practice_Code  \
0  THURROCK CCG  07G00         DR ABELA T PRACTICE        F81113   
1  THURROCK CCG  07G00  COMMONWEALTH HEALTH CENTRE        F81206   
2  THURROCK CCG  07G00  COMMONWEALTH HEALTH CENTRE        F81206   
3  THURROCK CCG  07G00        DR YADAVA N PRACTICE        F81211   
4  THURROCK CCG  07G00         DELL MEDICAL CENTRE        F81219   

          bnf_code                 id  hosp  \
0  0501070I0BCABAB   4658911000001108  True   
1  1501041T0AAABAB  27953211000001102  True   
2  1501041T0AAABAB  19711311000001103  True   
3  0501070I0AAABAB  17652311000001102  True   
4  0501021G0AAACAC  34751111000001106  True   

                            bnf_name  Items  Quantity  Actual_Cost  
0   Promixin_Inj 1,000,000u Vl (Dry)      1        56       195.22  
1   Midazolam HCl_Inj 2mg/ml 5ml Amp      1        10         6.66  
2   Midazolam HCl_Inj 2mg/ml 5ml Amp      1        10         6.66  
3  Colistimeth Sod_Inj 1,000,000u Vl      1        56        93.76  
4          Ceftriaxone Sod_Inj 1g Vl      1         3        26.71

In [7]:
#monthly total
df_ampp.sum()[['Items','Actual_Cost']]

Items             335685
Actual_Cost   3032677.99
dtype: object

In [8]:
##groupby chemical
df_ampp_product = df_ampp.groupby(['id', 'bnf_name', 'hosp']).sum().sort_values(by = 'Items', ascending = False)
df_ampp_product

Items  \
id                bnf_name                                 hosp           
17423911000001109 Laxido_Oral Pdr Sach (Orange) S/F        True  226420   
21495711000001101 Seebri_Breezhaler Inh Cap 55mcg + Dev    True   31820   
10844011000001107 Voltarol Emulgel_Aq Gel 1.16%            True   20748   
19711511000001109 Midazolam HCl_Inj 5mg/ml 2ml Amp         True   15617   
16614211000001109 Levest 150/30_Tab                        True   13164   
4840411000001107  Ventolin_Nebules Soln 2.5mg/2.5ml Ud     True    6490   
2154811000001100  Nexium_Tab 20mg                          True    3586   
4852411000001106  Isotretinoin_Cap 20mg                    True    3049   
1919011000001101  Kapake_Tab 30mg/500mg                    True    2721   
2155411000001101  Nexium_Tab 40mg                          True    2683   
4840611000001105  Ventolin_Nebules Soln 5mg/2.5ml Ud       True    2359   
2947311000001100  Estraderm MX 50_Patch                    True    2017   
8968311000001100  Pariet_Tab E/C 20mg                      True     952   
35204311000001101 SMA_PRO High Energy Liq                  True     658   
8968111000001102  Pariet_Tab E/C 10mg                      True     441   
17652311000001102 Colistimeth Sod_Inj 1,000,000u Vl        True     385   
9384411000001105  diabact UBT_Helicobacter Test Tab 50mg   True     202   
27953211000001102 Midazolam HCl_Inj 2mg/ml 5ml Amp         True     178   
19711311000001103 Midazolam HCl_Inj 2mg/ml 5ml Amp         True     178   
32975011000001108 Movicol_Ready to Take Liq Sach 25ml      True     140   
4851511000001102  Isotretinoin_Cap 5mg                     True     107   
24109111000001108 Carbagen_Tab 200mg                       True      96   
24109011000001107 Carbagen_Tab 200mg                       True      96   
22451511000001104 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
22786311000001102 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
23184711000001109 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
24920111000001108 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
34756111000001102 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
35751011000001102 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
23268811000001109 Tranexamic Acid_Inj 100mg/ml 5ml Amp     True      94   
...                                                                 ...   
15442911000001101 Piperacillin/Tazobactam_I/VInf 4g/0.5gVl True       3   
34750911000001102 Ceftriaxone Sod_Inj 250mg Vl             True       3   
23983711000001109 Gelaspan_Ecobag I/V Inf 4% 500ml         True       2   
8745311000001106  Ceftazidime Pentahyd_Inj 1g Vl           True       2   
4533011000001106  Cefotaxime Sod_Inj 500mg Vl              True       2   
34749911000001105 Cefotaxime Sod_Inj 1g Vl                 True       2   
34750511000001109 Ceftazidime Pentahyd_Inj 1g Vl           True       2   
34750111000001100 Cefotaxime Sod_Inj 500mg Vl              True       2   
34748511000001109 Amoxicillin Sod_Inj 1g Vl                True       2   
4682611000001109  Chirocaine_Inj 5mg/ml 10ml Amp           True       2   
3941011000001104  Cefotaxime Sod_Inj 1g Vl                 True       2   
3941111000001103  Cefotaxime Sod_Inj 1g Vl                 True       2   
10836511000001109 Nutricia_preOp Liq                       True       1   
34755311000001105 Piperacillin/Tazobactam_I/VInf2g/0.25gVl True       1   
12949711000001108 Tobramycin_Inj 40mg/ml 1ml Vl            True       1   
35543211000001108 Zoledronic Acid_I/V Inf 50mcg/ml 100mlVl True       1   
4438011000001106  Tazocin_I/V Inf 4g/0.5g Vl               True       1   
34752311000001104 Co-Amoxiclav_Inj 500mg/100mg Vl          True       1   
10530811000001109 Co-Amoxiclav_Inj 500mg/100mg Vl          True       1   
20372311000001100 Esomeprazole_Inj 40mg Vl (Dry)           True       1   
20552411000001106 Esomeprazole_Inj 40mg Vl (Dry)           True       1   
3475271100000110

A cursory look shows this AMPP list is very different from AMP list and appears to significant amount of non hospital Rx (e.g. laxido, Voltarol). The AMPP list consequently has a lot more items attrbuted to it. Explanations include
 - a) miscoded
 - b) AMPP does not neatly link to BNF code i.e. there is both a hosital and general pack available at AMPP which have both been mapped to a BNF code
 - c) Mapping provided by BSA is wrong

Example
17423911000001109	Laxido_Oral Pdr Sach (Orange) S/F	True	226420	2016364	1220049.09

The dmd code for this [product](https://openprescribing.net/dmd/ampp/17423911000001109/) which is a 4x2sachets pack size from Galen Ltd that has now been discontinued. It is possible that 
 -  this pack size was only available through hospitals
 -  b) above is correct and we can't accurately use this flag to map to prescribing
 - something else!